gpytorch/mlls/exact_marginal_log_likelihood.py

```diff
@@ -49,14 +60,7 @@ def forward(self, function_dist, target, *params):
         # Get the log prob of the marginal distribution
         output = self.likelihood(function_dist, *params)
         res = output.log_prob(target)
-
-        # Add additional terms (SGPR / learned inducing points, heteroskedastic likelihood models)
-        for added_loss_term in self.model.added_loss_terms():
-            res = res.add(added_loss_term.loss(*params))
-
-        # Add log probs of priors on the (functions of) parameters
-        for _, prior, closure, _ in self.named_priors():
-            res.add_(prior.log_prob(closure()).sum())
+        self._add_other_terms(res, params)
```

Comment:
I don't think this actually does anything (it returns the value but that return value is never assigned, this doesn't modify things in-place). 

Can awe instead make a `_compute_other_loss_terms` function that only takes in `params`, returns the additional value, and then add that explicitly in the forward method? That way things are less confusing.

Reply:
I meant to do the same as in `leave_one_out_pseudo_likelihood` where the returned value is assigned to `res`. Thanks for catching that. I have fixed that part.

Sure, we can do that, but what is the cleanest way for me to get the `dtype`, `device`, and shapes in that case since I'm not passing in res anymore? I can add three additional arguments to `_compute_other_loss_terms`, but maybe there is a better way of doing it.

Reply:
Hmm do you need that? I don't see these being used anywhere. Is there some implicit casting going on in `_add_other_terms` that you couldn't rely on otherwise? You could also have `_compute_other_loss_terms` be a generator that yields the individual terms, and then do `res = res + sum(self._compute_other_loss_terms(*params))` in the `forward).

Reply:
Is the way these losses are added up in the original code correct? If you have a batch-GP, then shouldn't the losses be added on independently for each model? The original code does stuff like this: `res.add_(prior.log_prob(closure()).sum())`, which would sum up the losses from all priors and then add them to each batch dimension, counting the losses multiple times when you eventually call `sum()` on `res` to do a backward pass?

Before changing this I need to understand what the shapes of these additional losses are. They will all have to be identical for us to call sum on them. Right now, they are all 0-dimensional tensors and it will be pretty easy to just sum them up, but it depends on if things are currently being handled correctly.

Reply:
Hmm yeah good point. I haven't been using batched priors with different values for different batch elements (though the code should allow for it). Same thing for the added loss term. Could this be generalized in a straightforward way if we just do `sum(dim=-1)` instead?

Reply:
That causes a lot of other tests to fail. As an example, the `test_train_on_batch_test_on_batch` test has four priors with sizes: `[2, 2]`, `[2]`, `[]`, [2, 1, 1]`, which obviously don't broadcast.

It's probably out of scope to fix this in this PR and it probably makes more sense to leave things as is and open a separate issue to track down how these additional losses are added up. What do you think?